tests/end_to_end/file_medium_size_write_then_read_client.cc

```diff
@@ -0,0 +1,140 @@
+void singleFileRead(const std::string& filename_base, ushort id) {
+  init();
+  const std::string filename(filename_base + std::to_string(id));
+  auto open_file_status(gfs::client::open(filename.c_str(),
+                                          gfs::OpenFlag::Read));
+  if (!open_file_status.ok()) {
+    LOG(ERROR) << "Open to read " + filename + " failed with error: "
+               << open_file_status;
+    exit(1);
+  } else {
+    LOG(INFO) << "Open to read " + filename + " succeeded";
+  }
+
+  auto read_data_or(gfs::client::read(
+      filename.c_str(), 0, kGlobalFileData.size()));
+
+  if (!read_data_or.ok()) {
+    LOG(ERROR) << "Read request in the " + std::to_string(id) 
+               << "-th thread failed due to: " << read_data_or.status();
+    exit(1);
+  } else {
+    LOG(INFO) << "Read request in the " + std::to_string(id) << " succeeded"; 
+  }
+
+  auto read_data(read_data_or.ValueOrDie());
+  if (memcmp(read_data.buffer, kGlobalFileData.c_str(), 
+             kGlobalFileData.size() != 0)) {
+    LOG(ERROR) << "Read request in the " + std::to_string(id) 
+               << " receives incorrect data";
+    exit(1);
+  } else {
+    LOG(INFO) << "Read request in the " + std::to_string(id)
+              << " receives correct data";
+  }
+}
+
+// Write a file and then read it from a different thread (so acting as a 
+// different client), verify the read
+void singleFileWriteThenRead(const std::string& filename_base, ushort id) {
+  init();
+  const std::string filename(filename_base + std::to_string(id));
+  auto create_file_status(gfs::client::open(filename.c_str(),
+                                            gfs::OpenFlag::Create));
+  if (!create_file_status.ok()) {
+    LOG(ERROR) << "Open to create " + filename + " failed with error: "
+               << create_file_status;
+    exit(1);
+  } else {
+    LOG(INFO) << "Open to create " + filename + " succeeded";
+  }
+
+  auto write_status(gfs::client::write(
+      filename.c_str(), (void*)kGlobalFileData.c_str(), 0, 
+      kGlobalFileData.size()));
+
+  if (!write_status.ok()) {
+    LOG(ERROR) << "Write request in the " + std::to_string(id) 
+               << "-th thread failed due to: " << write_status;
+    exit(1);
+  } else {
+    LOG(INFO) << "Write request in the " + std::to_string(id) << " succeeded"; 
+  }
+
+  // After successful write, spawn a reader thread and read to verify the write
+  std::thread reader_thread(singleFileRead, filename_base, id);
+  reader_thread.join();
+}
+
+// Launch a number of threads, and each thread first write a medium-size file 
+// (kGlobalFileData), and then spawn a separate thread to read the file and 
+// verify the data.
+void parallelFileWriteThenRead(const std::string& filename_base,
+                               const ushort num_of_threads) {
+  std::vector<std::thread> threads;
+  for (auto i = 0; i < num_of_threads; i++) {
+    threads.push_back(std::thread(singleFileWriteThenRead, filename_base, i));
```

Comment:
q: will it be beneficial that each thread actually read/write differenet data, so we know if GFS is doing things correctly

Reply:
Yes, ideally. I saved that because generating these data is expensive, but looking at the overall timing of this test I think we can do so.

Reply:
I made the data different among threads, looking good now.